src/mpid/ch4/shm/ipc/src/ipc_control.c

```diff
@@ -0,0 +1,128 @@
+/*
+ * Copyright (C) by Argonne National Laboratory
+ *     See COPYRIGHT in top-level directory
+ */
+
+#include "mpidimpl.h"
+#include "ipc_pre.h"
+#include "ipc_impl.h"
+#include "ipc_recv.h"
+#include "ipc_control.h"
+
+int MPIDI_IPC_send_lmt_fin_cb(MPIDI_SHM_ctrl_hdr_t * ctrl_hdr)
+{
+    int mpi_errno = MPI_SUCCESS;
+    MPIR_Request *sreq = (MPIR_Request *) ctrl_hdr->ipc_slmt_recv_fin.req_ptr;
+
+    MPIR_FUNC_VERBOSE_STATE_DECL(MPID_STATE_MPIDI_IPC_CTRL_SEND_LMT_RECV_FIN_CB);
+    MPIR_FUNC_VERBOSE_ENTER(MPID_STATE_MPIDI_IPC_CTRL_SEND_LMT_RECV_FIN_CB);
+
+    IPC_TRACE("send_lmt_recv_fin_cb: complete sreq %p\n", sreq);
+
+    MPIR_Datatype_release_if_not_builtin(MPIDIG_REQUEST(sreq, datatype));
+    MPID_Request_complete(sreq);
+
+  fn_exit:
+    MPIR_FUNC_VERBOSE_EXIT(MPID_STATE_MPIDI_IPC_CTRL_SEND_LMT_RECV_FIN_CB);
+    return mpi_errno;
+  fn_fail:
+    goto fn_exit;
+}
+
+int MPIDI_IPC_send_lmt_rts_cb(MPIDI_SHM_ctrl_hdr_t * ctrl_hdr)
```

Comment:
@minsii are you planning to make the ipc layer redirect to the right submodule callback based on `ipc_type`?

Reply:
The idea is that:
-  We provide generic single copy protocol (rts->single copy by receiver->fin) )in the IPC layer. The submodules should only provide memory handle related routines to utilize this protocol.
- For submodule custom protocols (e.g., coop copy based protocol in xpmem), the submodule define the header/callback by itself.

Reply:
If we find any submodule-custom protocol can be extracted as generic IPC protocol, we can move it to IPC layer in future. But it is unclear for now.

Reply:
I am replacing the old copy routines in the generic protocols with yaksa now. Then we can define the actual memory handle/cache API. Please let me know your thought on this.